gotestskip.go

```diff
@@ -114,27 +118,64 @@ func mainErr() error {
 			}
 			return fmt.Errorf("error decoding go test output: %v", err)
 		}
-		if skip.Packages[e.Package][e.Test] {
+		switch {
+		case shouldSkip(skip, e.Package, e.Test):
+			// The test is marked to be skipped.
 			switch e.Action {
 			case "fail":
 				e.Action = "skip"
 				wouldHaveFailed(e.Package, e.Test)
 			case "output":
-				e.Output = translateFailToSkip(e.Output)
+				translateFailOutput(&e, "SKIP")
+			case "run":
+				// The FAIL output for a parent test is produced
+				// before any of the FAIL (or Action: "fail") events
+				// of its subtests, so at that point we can't know
+				// whether to output FAIL or PASS.
+				// We could theoretically buffer output until the
+				// status of all subtests is known, but that's a bunch
+				// of complexity we don't want right now, so instead
+				// mark this test and its parents as skipping, so
+				// that we can delay the parent output until its status
+				// is known.
+				setSkipping(e.Package, e.Test)
+			}
+		case getStatus(e.Package, e.Test) == statusSuppressed:
+			// A subtest has failed but was skipped.
+			// If the test has failed, it's almost certainly because of this.
+			switch e.Action {
+			case "fail":
+				if e.Test == "" {
+					// Package-level status is marked as "success" not "pass".
+					e.Action = "success"
+				} else {
+					e.Action = "pass"
+				}
+			case "output":
+				translateFailOutput(&e, "PASS")
+			}
+			if err := sendSavedEvents(enc, e.Package, e.Test, "PASS"); err != nil {
+				return err
```

Comment:
Should this be set to PASS or SKIP if things failed, but we ignore the error?

I don't have a strong preference, just wondering what to set here.

If its set to skip, then tools like CircleCI would count this as skipped test and could be used as a metric.

Reply:
PASS, because we didn't ask to skip the test itself, and in general if a subtest is skipped, its parent will PASS not SKIP (there might be other subtests tests which were not skipped too).

This will only happen if some other test was skipped, so I think the metrics will still be useful.

Reply:
You are right. We're skipping a subtest, but passing parent test.